### ISSUE TITLE
youtube-dl -U == ERROR: can't find the current version. Please try again later.

### ISSUE THREAD

**Issue by jbrown123** [ytdl-org/youtube-dl#27003]
## Checklist

- [ ] I'm reporting a broken site support issue
- [x] I've verified that I'm running youtube-dl version **2020.11.12**
- [ ] I've checked that all provided URLs are alive and playable in a browser
- [ ] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-U', '-v']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2020.11.12
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: none
[debug] Proxy map: {}
Traceback (most recent call last):
  File "C:\Python\Python34\lib\urllib\request.py", line 1183, in do_open
  File "C:\Python\Python34\lib\http\client.py", line 1137, in request
  File "C:\Python\Python34\lib\http\client.py", line 1182, in _send_request
  File "C:\Python\Python34\lib\http\client.py", line 1133, in endheaders
  File "C:\Python\Python34\lib\http\client.py", line 963, in _send_output
  File "C:\Python\Python34\lib\http\client.py", line 898, in send
  File "C:\Python\Python34\lib\http\client.py", line 1287, in connect
  File "C:\Python\Python34\lib\ssl.py", line 362, in wrap_socket
  File "C:\Python\Python34\lib\ssl.py", line 580, in __init__
  File "C:\Python\Python34\lib\ssl.py", line 807, in do_handshake
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpwrmvn5tt\build\youtube_dl\update.py", line 46, in update_self
  File "C:\Python\Python34\lib\urllib\request.py", line 464, in open
  File "C:\Python\Python34\lib\urllib\request.py", line 482, in _open
  File "C:\Python\Python34\lib\urllib\request.py", line 442, in _call_chain
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpwrmvn5tt\build\youtube_dl\utils.py", line 2736, in https_open
  File "C:\Python\Python34\lib\urllib\request.py", line 1185, in do_open
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:600)>

ERROR: can't find the current version. Please try again later.
```

## Description

youtube-dl windows exe fails to update with -U flag.

Even after the github repo was restored this still fails. It appears from the debug trace that it is a SSL certificate issue. Interestingly if I manually go to the version URL https://yt-dl.org/update/LATEST_VERSION in my browser (latest windows chrome), I end up getting the 302 redirect, then I get the LATEST_VERSION file from github as expected. Likewise, if I use wget or curl from the command line I get the proper file with the current version. Only youtube-dl itself seems to have problems.

Perhaps an old version of the python SSL libraries is linked into the windows executable that doesn't support TLS or something related to the SSL connection handshake / setup?

```
c:\>sha256sum youtube-dl.exe
97a8c9ac735911a722b5405861ede8a66b02635f00664c5199bd01fd3b67114f *youtube-dl.exe

c:\>youtube-dl --version
2020.11.12

c:\>youtube-dl -U
ERROR: can't find the current version. Please try again later.
```

**Comment by MinkiNTN** [comment]
Have you try to use `youtube-dl -U --no-check-certificate` to update?

**Comment by jbrown123** [comment]
Just tried it. That at least got me the new version info. However I seem to have hung in the update process.

```
c:\>youtube-dl --version
2020.11.12

c:\>youtube-dl -U --no-check-certificate
Updating to version 2020.11.17 ...

c:\>Waiting for file handle to be closed ...
Updated youtube-dl to version 2020.11.17.
```

It doesn't return to the command prompt. I waited over 60 seconds with no response.
Hitting ENTER seemed to complete the update & return to the command prompt.

FYI: The update 2020.11.17 didn't fix the issue, but at least I know I can get the updates if I use `--no-check-certificate`

```
c:\>youtube-dl --version
2020.11.17

c:\>youtube-dl -U
ERROR: can't find the current version. Please try again later.

c:\>youtube-dl -U --no-check-certificate
youtube-dl is up-to-date (2020.11.17)
```

**Comment by Vangelis66** [comment]
@jbrown123 : I had no issue whatsoever updating from v**2020-11-12** to just released v**2020-11-17**: 
```
<redacted>\youtube-dl>youtube-dl -U -v
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-U', '-v']
[debug] Encodings: locale cp1253, fs mbcs, out cp737, pref cp1253
[debug] youtube-dl version 2020.11.12
[debug] Python version 3.4.4 (CPython) - Windows-Vista-6.0.6003-SP2
[debug] exe versions: ffmpeg N-97309-g4e0cf81b49, ffprobe N-97309-g4e0cf81b49, p
hantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
Updating to version 2020.11.17 ...
Waiting for file handle to be closed ...

<redacted>\youtube-dl>pause
Press any key to continue . . . Updated youtube-dl to version 2020.11.17.
```
My hunch is your issue is a duplicate of https://github.com/ytdl-org/youtube-dl/issues/25541

> yt-dl.org is sending the expired (30/05/2020) intermediate cert 
> (the last one in the below openssl output, "COMODO RSA Certification Authority"), 
> which is what's causing validation to fail in python. 
> Browsers are evidently smart enough to ignore it 
> and build an alternate chain, but the python client is not.

Possible duplicates are also https://github.com/ytdl-org/youtube-dl/issues/25469, https://github.com/ytdl-org/youtube-dl/issues/25476, https://github.com/ytdl-org/youtube-dl/issues/25491, https://github.com/ytdl-org/youtube-dl/issues/25536, https://github.com/ytdl-org/youtube-dl/issues/25537, https://github.com/ytdl-org/youtube-dl/issues/25539, https://github.com/ytdl-org/youtube-dl/issues/26168 and, possibly, others...

That issue is caused as a combination of the following factors:
1. The 32-bit `youtube-dl.exe` binary officially released is built on Python **3.4.4**, to satisfy Windows **XP** compatibility, at minimum (I'm not opposed to that, I myself am still on Vista SP2). 
2. The Windows (32-bit) version of Python **3.4.4** bundles OpenSSL v**1.0.2d** for secure connections.
3. py344+OpenSSL v**1.0.2d** has a bug, fixed in later versions: 

https://bugs.python.org/issue36011

4. During the check-for-available-update process, youtube-dl has to connect securely with the `yt-dl.org` domain; 
https://whatsmychaincert.com/?yt-dl.org => 

> yt-dl.org has a trusted chain containing an **expired** certificate. 
> This chain will work with modern web browsers 
> (they will use a chain to a different trusted root) 
> but may fail with older clients, notably OpenSSL 1.0.x.

5. On May 30th 2020, an intermediate server certificate issued by `yt-dl.org` expired; if that certificate is STILL present in your Windows CA store, then `youtube-dl.exe` (on py344) will use **that** instead of creating a new certificate chain with the server, and will thus fail the secure connection; related issue: https://github.com/ytdl-org/youtube-dl/issues/25622; however, another reason for staying at py344 is that the python (official) module (**py2exe**) used to (cross-)compile the Windows binary doesn't work on py3.5+ (but a [fork](https://github.com/albertosottile/py2exe/releases/) **does** exist...).

**TL;DR** => Possible issue solving scenario:
Navigate to your Windows CA store, by loading the `certmgr.msc` (certificates manager) snap-in; check on-line how to do that on your version of Windows: 

![crtmng](https://user-images.githubusercontent.com/9669492/99341158-063e1400-2892-11eb-92a8-ac48a5006d75.jpg)

Expand `Intermediate CAs` -> `Certificates`, then **sort stored intermediate certificates by Expiration Date**; then proceed to delete ALL the expired ones (but be careful)! Restart your machine, for good measure... If your Windows setup is fairly new, it might lack needed certificates, so try visiting first `yt-dl.org` with a (native) browser that uses the Windows store, like IE and/or Edge; the browser will then download needed updated certs and place them in the store, making them available to `youtube-dl.exe` for future connections.. 

Best wishes 😄

**Comment by pcislocked** [comment]
it was because github repo was down. now it should work

**Comment by Artanisx** [comment]
Still doesn't work without using the `--no-check-certificate` flag.

**Comment by jbrown123** [comment]
> it was because github repo was down. now it should work

That's not the problem. Note that in my original issue post I stated explicitly that `Even after the github repo was restored this still fails.`

**Comment by jbrown123** [comment]
@Vangelis66 Thank you for the very detailed explanation and solution.

I can confirm that it was the intermediate certificate problem. I deleted several COMODO certs and a couple of USERFIRST certs that all expired 5/30/2020 and the problem immediately went away.

I strongly encourage that this get posted to the FAQ as a solution for this specific error message.

**Comment by notela** [comment]
youtube-dl -U -v -4 --no-check-certificate

**Comment by jbrown123** [comment]
@notela Yes, the --no-check-certificate fixed it even before I deleted the expired intermediate certificates.

**Comment by ValZapod** [comment]
> I strongly encourage that this get posted to the FAQ as a solution for this specific error message.

This is not a fix. a fix is to update openssl.

**Comment by jbrown123** [comment]
Okay, call it a work around then. It's much simpler than explaining how to remove the expired intermediate certs or update the openssl version.

Thanks,

- James

**Comment by ValZapod** [comment]
You do understand that you cannot delete the certificate? Private key is still valid AND even not expired for code signing, and even after that if the binary is timestamped with secure server timestamp, thus signed but very old cert, the expiration of those AddTrust et al. certs will not mean expiration of signature?

**Comment by jbrown123** [comment]
Yes. It does. I did delete the expired intermediate keys and it solved the problem.

Use these instructions: https://github.com/ytdl-org/youtube-dl/issues/27003#issuecomment-728663509

Thanks,

- James

**Comment by Vangelis66** [comment]
@ValZapod wrote: 

> This is not a fix. a fix is to update openssl. (Python 3.8.2 has newer openssl.)

Well, it IS a fix in all intents and purposes, since the official devs team still wants to support **Windows XP** (py**344** last to work there) & **Windows Vista** (py**379** last to work there); py**38x** requires Win7 at minimum, plus the current stable python release, py39, requires Win8+; additionally, as I already pointed out, the team compiles the Windows executable via `py2exe` which, according to them, breaks in anything >py34x... 

If you so wish, you can use the latest Python release compatible with your specific Windows version and compile `youtube-dl.exe` via, e.g. `pyinstaller`; or install the desired py version system-wide and use the zipped-executable release, `youtube-dl`, or, even, install youtube-dl via pip python module; it isn't like there are no alternatives on Windows...

> You do understand that you cannot delete the certificate?

... of course you CAN; and while you're at it, one should also delete EXPIRED certificates inside `Trusted Roots Certification Authorities -> Certificates` (but also be extremely careful)!

> Private key is still valid AND even not expired for code signing, 
> and even after that if the binary is timestamped with secure server timestamp, 
> thus signed by very old (until 2030) cert, the expiration of those AddTrust et al. certs
> will not mean expiration of signature?

What I discussed in my first comment in this issue involves **intermediate server** certificates sent by remote servers while negotiating secure connections between the server and the python client inside `youtube-dl.exe`; these are cached by Windows in the Intermediate CAs certstore (to alleviate the need of re-fetching them from the web) and can, of course, **be removed at will**!

These intermediate server certs are ONLY relevant in so far a secure connection needs be established, they have nothing to do with code-signing Windows executables (`youtube-dl.exe` isn't code-signed, BTW...); I think you've probably mixed-up things here...

**Comment by jbrown123** [comment]
Thanks again @Vangelis66 for the detailed instructions and the FIX for this problem. While I think your answer (deleting the expired intermediate certificates) is the CORRECT answer, I also feel like the `--no-check-certificate` work-around is a good choice for those not comfortable with manually managing the certificate store.
It would be nice to include BOTH in the FAQ in my opinion.

**Comment by Vangelis66** [comment]
> I also feel like the `--no-check-certificate` work-around is a good choice

**Extra caution** should be exercised whenever using the `--no-check-certificate` switch to bypass a _broken certificate chain_, because in essence it **downgrades a secure (HTTPS) connection to a plain HTTP one**, to allow for the client to connect to the remote server at all cost; it should **only** be used when one is absolutely certain that the destination server has not been otherwise compromised... 

The CLI downloading tool `wget` has the same option, where associated **security risks** are explained in a more verbose fashion: 
https://www.gnu.org/software/wget/manual/html_node/HTTPS-_0028SSL_002fTLS_0029-Options.html

> --no-check-certificate
> 
> Don’t check the server certificate against the available certificate authorities.
> 
> As of Wget 1.10, the default is to verify the server’s certificate 
> against the recognized certificate authorities, breaking the SSL handshake 
> and aborting the download if the verification fails. 
> Although this provides more **secure** downloads, it does break interoperability 
> with some sites that worked with previous Wget versions, 
> particularly those using self-signed, **expired**, or otherwise invalid certificates. 
> **This option forces an “insecure” mode of operation 
> that turns the certificate verification errors into warnings and allows you to proceed.**
> 
> If you encounter “certificate verification” errors <snipped>, 
> you can use this option to bypass the verification and proceed with the download. 
> **Only use this option if you are otherwise convinced of the site’s authenticity, or 
> if you really don’t care about the validity of its certificate. 
> It is almost always a bad idea not to check the certificates 
> when transmitting confidential or important data.**

IOW, though this is an extremely unlikely scenario (but not unheard off...), imagine the case where `yt-dl.org` has been hacked; then the malicious actor (MITM) could redirect your `--no-check-certificate` **insecure** "check-for-update" request to somewhere else and/or serve you malicious code...

Just offered as food-for-thought...

**Comment by jbrown123** [comment]
I agree completely. I just suggest it as a simpler alternative than having people root around in their certificate store. Many people aren't comfortable doing that and there's a possibility of doing "damage" to your system that is difficult (though clearly not impossible) to recover from.

I think both should be offered with appropriate warnings. If not, I'm good with just telling people how to remove the offending certificates.

Thanks,

- James

**Comment by ValZapod** [comment]
> of course you CAN; and while you're at it, one should also delete EXPIRED certificates

No, you must not, like Microsft did not, it is illegal from standard point of view to remove expired root certificates, and that private key did not even expire yet until 2023. https://crt.sh/?id=162879063

You know nothing about global root PKI structure, are you? For example, VISA root keys and others? As an expert, I can point you to RFC 4158 that says that you MUST verify even expired chains and to check whether there is at least one that is not expired. How will you be able to do it if you delete one of expired root certs? This is funny. In this case we have one of intermediates that have root version, but still.

**Comment by Vangelis66** [comment]
As things stand, I clearly state that I'm **NOT** an authority on certification chains, unlike my dear friend **Valerii** 😄 ; but I definitely state that when **that same issue** hit me in the start of June 2020, I, as a mere youtube-dl user on Windows (Vista SP2 32-bit), have researched it quite a lot... 

Yes, down the line, **it is a genuine bug** of the old (1.0.2d) openssl lib bundled with CPython 3.4.4 that is used by the devs to package the official Windows binary; yes, that "bug" is manifested in the case discussed due to the `yt-dl.org` server having sent - in an era prior to May 30th 2020 - a now expired **intermediate** certificate, which had been cached in the CA certstores of the affected machines. FWIW, admins of that server should have reconfigured it after 30/05/2020, to remove the expired certificate "out of the picture"... 

Yes, newer versions of openssl and recent browsers will disregard the (eventual) presence of that expired intermediate cert and fallback to building an alternative, **valid**, chain; so I don't see why Chrome is again brought into this discussion... Yes, updated browsers WORK; this issue, though, is not about a browser...

**Point 1**: The chief youtube-dl dev himself points out in https://github.com/ytdl-org/youtube-dl/issues/25491#issuecomment-637047754 : 

> yt-dl.org certificate is fine and not expired. 
You're hit by expired AddTrust External CA Root 
**which is still on your CA store**

Further comments by users on that same issue, #25491, are indeed quite spot on...

**Point 2**: The many-a-times-already linked article in 
https://www.agwa.name/blog/post/fixing_the_addtrust_root_expiration
suggests that the cure for the issue is **removal** of the expired **AddTrust External CA Root** cert from the CA store - on _Debian_; why isn't that applicable on Windows?

**Point 3**: My offered solution in https://github.com/ytdl-org/youtube-dl/issues/27003#issuecomment-728663509 is just a modified (for `yt-dl.org`) iteration of the instructions under https://github.com/ytdl-org/youtube-dl/issues/7309#issuecomment-172354527 😉 ; but just to be _on the safe side_, I only suggested the removal of expired certs from the **Intermediate Certification Authorities** store (not from **Trusted Root**/**Third-Party Root** CAs, where I do still have the [expired] **AddTrust External CA Root**...). 

**Point 4**: For anyone having lots of spare time, he/she can further inspect the log I had saved at the time (June 11th, 2020) by running: 
https://www.ssllabs.com/ssltest/analyze.html?d=yt%2ddl.org&s=95.143.172.170&latest
=> Log: [CertificationPaths.txt](https://github.com/ytdl-org/youtube-dl/files/5578849/CertificationPaths.txt)
With the **COMODO RSA Certification Authority** _intermediate_ cert (issued by now expired **AddTrust External CA Root**) present, `youtube-dl.exe` was, by default, attempting **Path No. 2**, then aborting the connection attempt. After I had **removed** it from the Intermediate CAs store, then secure connection to `yt-dl.org` was successful (via either Path No.1 or No.3, wasn't sure...). For the record, I haven't witnessed any malfunction on my system in these 5 1/2 months after the removal of said expired cert (but why should you take my word for it? 😆 ) ... 
As for the _legality_ of removing expired certs, **intermediate** and/or **root** ones, I totally lack the knowledge to comment upon... 

All-in-all, I don't quite see why I'm being _selectively_ picked by @ValZapod, where many other people in many other places simply suggest a "solution" similar to mine... In any case, I won't be back for a further rebuttal, I consider this specific issue "closed" as far as I'm concerned...

@dstftw, do you have anything concrete to add to this discussion?

@jbrown123: If you feel your original **issue** has been successfully mitigated, perhaps now is the time to **close** it?

Kind regards to all, please stay safe during these perilous times... 😅

**Comment by ValZapod** [comment]
Who was right? I was right. You should never delete expired root certs. Official Microsoft recommendation. https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/trusted-root-certificates-are-required

**Comment by jbrown123** [comment]
If it's that important to you, sure you can be "right".

Now, for the rest of us ... if you need your youtube-dl executable to work on windows, deleting expired certificates is a completely safe and functional work around.

**Comment by goodboi69** [comment]
I am currently experiencing this issue as well.

**Comment by scrutinizer11** [comment]
I run **youtube-dl** on MacOS, **python 2.7.16**. I second the OP and everyone who encountered this error. Deleting every expired **_Intermediate_** certificate did not cure it. I can't run the update without `--no-check-certificate`.
